### PR TITLE
fix(healthcheck): fix readiness ticker resetting sampling window on recovery

### PR DESCRIPTION
**Pull request by OriginLeon**
## Summary

Fixes ROUTER-1618 (escalation RH-1297).

The readiness ticker had two bugs:

- **Sampling window reset on recovery**: `interval_at` was created inside the outer loop, so the sampling clock restarted from zero on every recovery cycle, causing inconsistent window behavior.
- **Rejection counter race**: the counter was read and zeroed with separate `load` + `store(0)` operations, leaving a window where rejections arriving between the two could be silently dropped.

**Changes:**
- Move `interval_at` outside the loop so the sampling clock is stable across recovery cycles
- Set `MissedTickBehavior::Delay` to prevent catch-up ticks after the recovery sleep
- Use `swap(0, Relaxed)` to atomically read and reset the rejection counter

**Tests added:**
- Recovery: verifies the router returns to UP after the recovery interval (DOWN→UP)
- Multiple unready cycles: verifies the counter resets correctly between cycles, allowing the router to go unready again after recovering
- Boundary conditions: exactly `allowed` rejections stays UP; `allowed + 1` goes DOWN

## Test plan

- [x] `cargo test -p apollo-router healthcheck` — 11/11 passed
- [x] `cargo clippy -p apollo-router -- -D warnings` — clean
- [x] `cargo fmt` — no changes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apollo-router/src/plugins/healthcheck/mod.rs [hunk 1]
```diff
@@ -200,21 +200,22 @@ impl PluginPrivate for HealthCheck {
         let my_ready = ready.clone();
 
         let ticker = tokio::spawn(async move {
+            let start = Instant::now() + my_sampling_interval;
+            let mut sampling_interval = tokio::time::interval_at(start, my_sampling_interval);
```
Comment (carodewig) [inline apollo-router/src/plugins/healthcheck/mod.rs]: Nit: you can just use `tokio::time::interval` now - there's no real benefit to waiting until after the first tick since there won't be any rejections in it

File: apollo-router/src/plugins/healthcheck/mod.rs [hunk 2]
```diff
@@ -200,21 +200,22 @@ impl PluginPrivate for HealthCheck {
...
-                        my_ready.store(true, Ordering::SeqCst);
-                        break;
-                    }
+                sampling_interval.tick().await;
+                let rejected_count = my_rejected.swap(0, Ordering::Relaxed);
+
+                if rejected_count > allowed {
+                    my_ready.store(false, Ordering::SeqCst);
+                    tokio::time::sleep(my_recovery_interval).await;
+                    my_ready.store(true, Ordering::SeqCst);
```
Comment (carodewig) [inline apollo-router/src/plugins/healthcheck/mod.rs:213]: I think you need to store 0 here again, or else take an approach that performs sampling in one loop and stores readiness in another (ie via a channel).

As written, this will sample rejections over the duration of `my_recovery_interval` which defaults to being longer than the sampling interval, so you could end up marking the pod as unready when it is fine.

File: apollo-router/src/plugins/healthcheck/mod.rs [hunk 3]
```diff
@@ -540,4 +541,236 @@ mod test {
...
+    }
+
+    // Helper to assert the JSON health body and HTTP status of a response
+    async fn assert_health_response(
+        response: http::Response<axum::body::Body>,
+        expected_status: StatusCode,
+        expected_json: &str,
+    ) {
+        assert_eq!(expected_status, response.status());
+        let j: serde_json::Value = serde_json::from_slice(
```
Comment (carodewig) [inline apollo-router/src/plugins/healthcheck/mod.rs]: Nit: use a descriptive variable name rather than `j`

File: apollo-router/src/plugins/healthcheck/mod.rs [hunk 4]
```diff
@@ -540,4 +541,236 @@ mod test {
...
+            &crate::services::router::body::into_bytes(response)
+                .await
+                .expect("body"),
```
Comment (carodewig) [inline apollo-router/src/plugins/healthcheck/mod.rs]: Nit: expect messages should contain the detail for what should have happened (ie 'body should be parseable')

File: apollo-router/src/plugins/healthcheck/mod.rs (hunk 2 above)
Comment (OriginLeon) [reply apollo-router/src/plugins/healthcheck/mod.rs:213]: Thank for that you are correct. Working on fix.

File: apollo-router/src/plugins/healthcheck/mod.rs (hunk 3 above)
Comment (OriginLeon) [reply apollo-router/src/plugins/healthcheck/mod.rs]: Sorry about that forgot to do some cleaning up. Fixed

File: apollo-router/src/plugins/healthcheck/mod.rs (hunk 4 above)
Comment (OriginLeon) [reply apollo-router/src/plugins/healthcheck/mod.rs]: I have added a more descriptive message

File: apollo-router/src/plugins/healthcheck/mod.rs (hunk 1 above)
Comment (OriginLeon) [reply apollo-router/src/plugins/healthcheck/mod.rs]: good catch, switched to interval(). Had to bump the recovery test to use the 5s config since the first tick now fires during setup, which shifts when the second tick lands relative to the test assertions.
